File: Algorithms/Graphs/Breadth First Search/Java/BreadthFirstSearch.java
/* Java implementation of BFS algorithm to traverse a graph.
 *
 * Starting from the root node, this algorithm explores all the neighboring nodes
 * and keeps track of visited nodes to make sure no node is visited more than once. */

import java.util.Iterator;
import java.util.LinkedList;

public class BreadthFirstSearch {

    /* Number of vertices */
    private int V;
    /*Adjacency Lists */
    private LinkedList<Integer> adj[];

    BreadthFirstSearch(int v)
    {
        V = v;
        adj = new LinkedList[v];
        for (int i = 0; i < v; ++i)
            adj[i] = new LinkedList();
    }

    /* Add an edge to the graph. Node w is adjacent to given Node v */
    void addEdge(int v,int w)
    {
        adj[v].add(w);
    }

    /* BFS algorithm which starts from a given node, s */
    void BFS(int s)
    {
        /* List of visited vertices; all false in the beginning) */
        boolean visited[] = new boolean[V];

        /* Queue data structure is used for BFS */
        LinkedList<Integer> queue = new LinkedList<>();

        /* Mark starting node s as visited and enqueue it */
        visited[s]=true;
        queue.add(s);

        /* Until queue is empty, dequeue a single node in queue and look for it's neighboring vertices.
         *  If an adjecent node hasn't been visited yet, set it as visited and enqueue this node. s*/
        while (queue.size() != 0)
        {
            /* Dequeue */
            s = queue.poll();
            System.out.print( s + " ");

            /* Get all adjacent vertices */
            Iterator<Integer> i = adj[s].listIterator();
            while (i.hasNext())
            {
                int n = i.next();
                if (!visited[n])
                {
                    visited[n] = true;
                    queue.add(n);
                }
            }
        }
    }

    public static void main(String[] args) {

        /* Create Graph with 4 vertices */
        BreadthFirstSearch g = new BreadthFirstSearch(4);

        /* Add edges in graph */
        g.addEdge(0, 0);
        g.addEdge(0,1);
        g.addEdge(0, 2);
        g.addEdge(1, 2);
        g.addEdge(2, 0);
        g.addEdge(2, 3);

        System.out.println("Traverse graph with BFS...");
        System.out.println("Starting Vertex: 2");

        g.BFS(2);
    }
}

File: Algorithms/Searching & Sorting/Breadth-First-Search.cs
// Source - https://www.geeksforgeeks.org/breadth-first-search-or-bfs-for-a-graph/
// C# program to print BFS traversal 
// from a given source vertex. 
// BFS(int s) traverses vertices 
// reachable from s. 
using System; 
using System.Collections.Generic; 
using System.Linq; 
using System.Text; 

// This class represents a directed 
// graph using adjacency list 
// representation 
class Graph{ 
	
// No. of vertices	 
private int _V; 

//Adjacency Lists 
LinkedList<int>[] _adj; 

public Graph(int V) 
{ 
	_adj = new LinkedList<int>[V]; 
	for(int i = 0; i < _adj.Length; i++) 
	{ 
		_adj[i] = new LinkedList<int>(); 
	} 
	_V = V; 
} 

// Function to add an edge into the graph 
public void AddEdge(int v, int w) 
{		 
	_adj[v].AddLast(w); 

} 

// Prints BFS traversal from a given source s 
public void BFS(int s) 
{ 
	
	// Mark all the vertices as not 
	// visited(By default set as false) 
	bool[] visited = new bool[_V]; 
	for(int i = 0; i < _V; i++) 
		visited[i] = false; 
	
	// Create a queue for BFS 
	LinkedList<int> queue = new LinkedList<int>(); 
	
	// Mark the current node as 
	// visited and enqueue it 
	visited[s] = true; 
	queue.AddLast(s);		 

	while(queue.Any()) 
	{ 
		
		// Dequeue a vertex from queue 
		// and print it 
		s = queue.First(); 
		Console.Write(s + " " ); 
		queue.RemoveFirst(); 
		
		// Get all adjacent vertices of the 
		// dequeued vertex s. If a adjacent 
		// has not been visited, then mark it 
		// visited and enqueue it 
		LinkedList<int> list = _adj[s]; 

		foreach (var val in list)			 
		{ 
			if (!visited[val]) 
			{ 
				visited[val] = true; 
				queue.AddLast(val); 
			} 
		} 
	} 
} 

// Driver code 
static void Main(string[] args) 
{ 
	Graph g = new Graph(4); 
	
	g.AddEdge(0, 1); 
	g.AddEdge(0, 2); 
	g.AddEdge(1, 2); 
	g.AddEdge(2, 0); 
	g.AddEdge(2, 3); 
	g.AddEdge(3, 3); 
	
	Console.Write("Following is Breadth First " + 
				"Traversal(starting from " + 
				"vertex 2)\n"); 
	g.BFS(2); 
} 
} 
